### ISSUE TITLE
[PATCH] twisted.xish.domish.elementStream fails to parse xml

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11251]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| amberite reported|
|-|-|
|Trac ID|trac#530|
|Type|defect|
|Created|2004-03-01 03:24:05Z|

Attachments:

* [xish.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c6/5c64cc5d7cc05237d16fa2873e58b5923c489b73/893882a46de229ee47479c4b1b3efe0c4d46bbc3.patch) (3308 bytes) - added by ralphm on 2005-03-04 19:11:32Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__530 530
type__defect defect
reporter__amberite amberite
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__words words
time__1078111445000000 1078111445000000
changetime__1472370309036739 1472370309036739
version__ 
owner__ 
cc__itamarst cc__dizzyd cc__amberite cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
Twisted 1.2 on MacOS 10.3.2

The following xml cannot be parsed, using either SuxElementStream or
ExpatElementStream:

&lt;?xml version="1.0" encoding="UTF-8"?>
&lt;root>
    &lt;category cat-id="A_Z" cat-name="A-Z", doc-count="33"/>
&lt;/root>

SuxElementStream produces this traceback:

Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
  File
"/System/Library/Frameworks/Python.framework/Versions/2.3/lib/python2.3/site-packages/twisted/protocols/sux.py",
line 180, in dataReceived
    newState = doFn(byte)
  File
"/System/Library/Frameworks/Python.framework/Versions/2.3/lib/python2.3/site-packages/twisted/protocols/sux.py",
line 238, in do_tagstart
    self.gotTagStart(self.tagName, {})
  File
"/System/Library/Frameworks/Python.framework/Versions/2.3/lib/python2.3/site-packages/twisted/xish/domish.py",
line 506, in gotTagStart
    self.DocumentStartEvent(e)
TypeError: 'NoneType' object is not callable

ExpatElementStream produces a different traceback:

Traceback (most recent call last):
  File "&lt;stdin>", line 1, in ?
  File
"/System/Library/Frameworks/Python.framework/Versions/2.3/lib/python2.3/site-packages/twisted/xish/domish.py",
line 593, in parse
    self.parser.Parse(buffer)
  File
"/System/Library/Frameworks/Python.framework/Versions/2.3/lib/python2.3/site-packages/twisted/xish/domish.py",
line 606, in _onStartElement
    e = Element(qname, self.defaultNsStack[-1], attrs)
IndexError: list index out of range
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
&lt;category cat-id="A_Z" cat-name="A-Z", doc-count="33"/>
                                     ^
                               this comma is illegal

Possibly the error messages could be better, but that is obviously not legal XML.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
My apologies, the comma you pointed out was a typo on my part when entering the bug report.  In my 
test case, that comma does not exist (the XML is valid), but the tracebacks occur as described in the 
original note.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
The attached patch fixes my issue with elementStream, at least when using the
ExpatElementStream implementation.  I'll try to follow up with a patch to fix
SuxElementStream as well, once I understand the code better.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
Assigning this to itarmarst, since he was the person who responded initially.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I also suspect '-' can't be used in XML attributes. A brief look at the spec
indicates I am correct, but better check yourself.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
Sorry, I'm really not trying to be a pest, but the hyphens are legal (see
below), and the bug I am reporting does exist (see my initial patch, which
partially fixes this).

Seriously, domish.elementStream can't even parse something as simple as
'''&lt;foo>&lt;/foo>''' at the moment.  My patch at least fixes the
ExpatElementStream, which is better than leaving domish broken.

I took your advice and did my homework on the XML specification, and www.w3.org
defines an attribute name as follows:

AttDef ::= S Name S AttType S DefaultDecl

where "Name" is the name of the attribute, and "Name" is defined as:

Name ::= (Letter | '_' | ':') (NameChar)*

where NameChar is defined as:

NameChar ::= Letter | Digit | '.' | '-' | '_' | ':' | CombiningChar | Extender
                                   ^^^^^
                              Underscore is legal

Again, sorry to be a pest, but I'd very interested in seeing this fixed, and am
willing to rework the submitted patch until it meets your requirements.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
You're not being a pest, you're being helpful. I'm assigning this to dizzyd, who
is the xish maintainer.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dizzyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dizzyd commented|
|-|-|

```
#!html
<pre>
This bug actually has absolutely nothing to do with attributes, and everything to do with 
a lack of namespaces on the example document passed into the ElementStreams. To be honest, I've 
never actually used XML without namespaces, hence the lack of support for missing/empty/non-
existent namespaces in domish.

If this is a real deal breaker, I'll consider adding support for missing/empty/non-existent namespaces, 
but you'll still have to deal with QNames and such -- they'll just have no URI.

Comments?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
Thanks for the explanation.  My experience is the exact opposite of yours, in
that none of the xml I interact with at my place of business *ever* uses namespaces.

My particular use case involves interacting with a commercial search engine,
which produces search results in an xml format similar to the simplified example
I provided in the initial bug report.  Since the requirements for
well-formedness do not mandate the use of namespaces within an xml document, and
since I'm dealing with a commercial product, not being able to handle this kind
of xml data is a deal breaker for my use of twisted.

Generating an element as follows would be perfectly fine by me:

e = domish.Element((None, 'result'), attribs={'cat-id':'Pharma', 'doc-count':33})

The patch I provided allows for this, although I don't know if it does things
"the right way" for twisted.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dizzyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dizzyd commented|
|-|-|

```
#!html
<pre>
Fair enough -- I'll make the appropriate changes this evening to domish.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

```
#!html
<pre>
These changes don't seem to have taken place. Is this still current?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm commented|
|-|-|

```
#!html
<pre>
The attached patch xish.patch fixes the issue, but simpler. Also added two test
cases. The patch depends on the renaming of xish.xpath's intern to internQuery
(see patch attached to #913).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dizzyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dizzyd commented|
|-|-|

```
#!html
<pre>
Had to adjust xish.patch a little, but merged it into SVN.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_16"></a>|@markrwilliams commented|
|-|-|

In changeset fb29a1cc14c15203e0f64ecc99e5225ed753e615
```
#!CommitTicketReference repository="" revision="fb29a1cc14c15203e0f64ecc99e5225ed753e615"
Merge pull request #530 from twisted/8793-pb-newstyletests-iopump

Merge 8793-pb-newstyletests-iopump: Make Perspective Broker's NewStyleTests use IOPump

Author: markrwilliams
Reviewer: rodrigc
Fixes: #8793

Introduce twisted.spread.test.test_pb.connectServerAndClient, which takes a PBClientFactory and PBServerFactory and connects the objects obtained from the two via an IOPump.  NewStyleTests now uses this instead of a TCP connection over the loopback interface.
```
